// ngx-markdown-editor/esm2020/public_api.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { MarkdownEditorComponent } from './lib/md-editor.component';
import { MarkdownEditorResizeSensorComponent } from './lib/resize-sensor/resize-sensor.component';
import * as i0 from "@angular/core";
export { MarkdownEditorComponent } from './lib/md-editor.component';
export { MarkdownEditorResizeSensorComponent } from './lib/resize-sensor/resize-sensor.component';
export class LMarkdownEditorModule {
}
LMarkdownEditorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: LMarkdownEditorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LMarkdownEditorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: LMarkdownEditorModule, declarations: [MarkdownEditorComponent,
        MarkdownEditorResizeSensorComponent], imports: [CommonModule,
        FormsModule], exports: [MarkdownEditorComponent,
        MarkdownEditorResizeSensorComponent] });
LMarkdownEditorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: LMarkdownEditorModule, imports: [[
            CommonModule,
            FormsModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: LMarkdownEditorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        MarkdownEditorComponent,
                        MarkdownEditorResizeSensorComponent
                    ],
                    imports: [
                        CommonModule,
                        FormsModule
                    ],
                    exports: [
                        MarkdownEditorComponent,
                        MarkdownEditorResizeSensorComponent
                    ]
                }]
        }] });
