// ngx-markdown-editor/esm2020/lib/resize-sensor/resize-sensor.component.mjs
import { Component, Input, Output, ViewChild, EventEmitter } from '@angular/core';
import * as i0 from "@angular/core";
export class MarkdownEditorResizeSensorComponent {
    constructor(_ngZone) {
        this._ngZone = _ngZone;
        this.interval = 500;
        this.resize = new EventEmitter();
        this.sizeInfo = {
            width: 0,
            height: 0
        };
    }
    ngAfterViewInit() {
        this.sizeInfo = {
            width: this.resizeSensor.nativeElement.offsetWidth,
            height: this.resizeSensor.nativeElement.offsetHeight
        };
        this._ngZone.runOutsideAngular(() => {
            setTimeout(() => {
                this.detectSize();
            }, this.interval);
        });
    }
    detectSize() {
        let width = this.resizeSensor.nativeElement.offsetWidth;
        let height = this.resizeSensor.nativeElement.offsetHeight;
        if (this.sizeInfo.width !== width || this.sizeInfo.height !== height) {
            this.sizeInfo = {
                width: width,
                height: height
            };
            this.resize.emit(this.sizeInfo);
        }
        setTimeout(() => {
            this.detectSize();
        }, this.interval);
    }
}
MarkdownEditorResizeSensorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: MarkdownEditorResizeSensorComponent, deps: [{ token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Component });
MarkdownEditorResizeSensorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.1", type: MarkdownEditorResizeSensorComponent, selector: "md-editor-resize-sensor", inputs: { interval: "interval" }, outputs: { resize: "resize" }, viewQueries: [{ propertyName: "resizeSensor", first: true, predicate: ["resizeSensor"], descendants: true, static: true }], ngImport: i0, template: "<div class=\"md-editor-resize-sensor\" #resizeSensor></div>\r\n", styles: [".md-editor-resize-sensor{position:absolute;left:0;right:0;top:0;bottom:0;z-index:-1}\n"] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: MarkdownEditorResizeSensorComponent, decorators: [{
            type: Component,
            args: [{ selector: 'md-editor-resize-sensor', template: "<div class=\"md-editor-resize-sensor\" #resizeSensor></div>\r\n", styles: [".md-editor-resize-sensor{position:absolute;left:0;right:0;top:0;bottom:0;z-index:-1}\n"] }]
        }], ctorParameters: function () { return [{ type: i0.NgZone }]; }, propDecorators: { resizeSensor: [{
                type: ViewChild,
                args: ['resizeSensor', { static: true }]
            }], interval: [{
                type: Input
            }], resize: [{
                type: Output
            }] } });
